### DISCUSSION TITLE
Unable to change chains

### DISCUSSION THREAD

**Discussion by Dilosen** [WalletConnect/WalletConnectUnity#94] [Q&A] [answered]
Hello All

I am unable to change the default chains in the demo. I have set up a project in the walletconnect dashboard and put the id into the respective field in the walletconnectUnity prefab - however the list seems to be the default list no matter what I try. I can't seem to find any other way, I have submitted my app for review, however it says that I should be able to test the SDKs with the project ID. 

Not sure what to do from here as the chain I need is not displaying within the default list.

**Comment by Dilosen** [comment]
I am specifically looking to use the algorand chain, I have tried to add the chain to the chain script, but I am getting an error that the dapp maybe in testnet and the wallet in mainnet etc. - The EVM ones work, but my hack of it doesn't seem too want to work. 

I am hoping that there will be some intergration for algorand - I got the ID's from the 'Explorer' section of the WalletConnect dashboard, but I am not sure if these are correct.

**Comment by sebas6612** [comment]
Hi @moosesnWoop  I'm not part of the Wallet Connect team but I'm handling the same situation, my solution was:

1. add the prefab
2. In the Unity editor assign the prefab to the `WCSignClient _client;`
3. set up the configuration for Algorand main net: (the most important part and probably were you are having the issues): 

```
var requiredNamespaces = new RequiredNamespaces().WithProposedNamespace(
                "algorand", new ProposedNamespace()
                {
                    Methods = new[]
                    {
                        "algo_signTxn", 
                    },
                    Chains = new[]
                    {
                        "algorand:wGHE2Pwdvd7S12BL5FaOP20EGYesN73k" //Algorand Mainnet
                    },
                    Events = new[]
                    {
                        "accountChanged",
                    }
                });
```


**_The full code bellow:_**
```
using System;
using System.Collections.Generic;
using System.Linq;
using UnityEngine;
using WalletConnectSharp.Sign.Models;
using WalletConnectSharp.Sign.Models.Engine;
using WalletConnectUnity.Utils;
using WalletConnect;
using System.Threading.Tasks;
using UnityEngine.UI;


namespace Assets.Scripts.UI.Demo.WCController
{
    public class WCController : MonoBehaviour
    {
        [SerializeField] private WCSignClient _client;

        public Button ConnectBtn;

        void Start()
        {
            ConnectBtn.onClick.AddListener(OnClick);
        }

        private async void OnClick()
        {
            // Call Async Function
            await ExecuteWCConnection();
        }

        private async Task ExecuteWCConnection()
        {
            if (_client == null)
            {
                Debug.LogError("No WCSignClient scripts found in scene!");
                return;
            }

            if (_client.SignClient == null)
                await _client.InitSignClient();

            // Connect Sign Client
            Debug.Log("Connecting sign client..");

            var requiredNamespaces = new RequiredNamespaces().WithProposedNamespace(
                "algorand", new ProposedNamespace()
                {
                    Methods = new[]
                    {
                        "algo_signTxn",
                    },
                    Chains = new[]
                    {
                        "algorand:wGHE2Pwdvd7S12BL5FaOP20EGYesN73k"
                    },
                    Events = new[]
                    {
                        "accountChanged",
                    }
                });

            var dappConnectOptions = new ConnectOptions()
            {
                RequiredNamespaces = requiredNamespaces
            };

            var connectData = await _client.Connect(dappConnectOptions);

            Debug.Log($"Connection successful, URI: {connectData.Uri}");

            try
            {
                await connectData.Approval;  // IMPORTANT: the code only gets to this line and awaits indefinitely even after pressing the connect button in the wallet app

                // We need to move this to the main unity thread
                // TODO Perhaps ensure we are using Unity's Sync context inside WalletConnectSharp
                MTQ.Enqueue(() =>
                {
                    Debug.Log($"Algorand: {connectData.Approval}");
                    Debug.Log($"Algorand2: {connectData.Approval}");
                    Debug.Log($"Algorand3: {connectData.Approval.Result.Key}");
                    Debug.Log($"Algorand4: {connectData.Approval.Result}");
                    Debug.Log($"Connection approved, URI: {connectData.Uri}");
                });
            }
            catch (Exception e)
            {
                Debug.LogError(("Connection failed: " + e.Message));
                Debug.LogError(e);
            }
        }
    }
}

```

Hope this is helpful and if you find a better way please let me know

**Comment by Dilosen** [comment]
Thanks for your help Brother, I actually gave up on this last week, hoping that Algorand/Unity would get themselves sorted out. 
This is great and I will try implemenation tonight. 

May I ask where you got the namespace for the Algorand Chains? (    "algorand:wGHE2Pwdvd7S12BL5FaOP20EGYesN73k")

Also, I learnt from the Unity/Algorand Devs that WalletConnect isn't even required, that you can make the game the wallet itself and process txn within the game.  Alas, that was because WalletConnect wasn't working for us at the time. 

Looking forward to testing this code out

**Reply by sebas6612** [reply]
Hi @moosesnWoop 
I agree with you, Algoran should have a sample implementation ready to facilitate this process.

Regarding your question, I had to investigate and I understood that "WalletConnect" follows a standard called "Chain agnostic", here is the link to [Algoran namespaces](https://github.com/ChainAgnostic/namespaces/blob/main/algorand/caip2.md).

Also, yes you are right you can make the game the wallet itself, but I think it is better to use WC and let big wallets handle the security concerns, and as game developers just focus on the Game.

Finally just to mention the example I gave to you is using WC V2 but I see they just released WC V3 publicly available a few hours ago, so supposed to be easier now.

**Reply by Dilosen** [reply]
Hi @sebas6612 - I have implemented the new packages for the UnityWalletConnect, specifically the modal - which I think is the perfect thing I need to implement. 

I'm now working through this with your Algorand proposedNameSpace instead to see how I can get it up and running.

For now there is no "game" while I try to get this thing to work

**Reply by sebas6612** [reply]
Yes, the modal is the right way but I haven't had the time to check the new version, but soon I will try it. For the moment if you find a way to make it work please let me know

**Reply by Dilosen** [reply]
I'm looking through it now.  Looks like we can implement Algorand Chain and add it to the Chain.cs, but I need to look into the code today to make sure

There is a working modal sample scene - 

1)update the packages in the package manager
2) on the modal package, go to samples tab
3) import samples

There's now a working scene to look through - going to get some coffee!

**Reply by skibitsky** [reply]
Hey @moosesnWoop, if you add the Algorand Chain to Chain.cs, please share the code. I will then include it in our sample to make it easier for other Algorand developers.

**Reply by sebas6612** [reply, marked as answer]
Hi @skibitsky the WC V3 is working great, I made it work for Algorand in my game, and is doing perfectly with [Pera Algo Wallet](https://perawallet.app/) but not working with [Defly Wallet](https://defly.app/) (I assume they haven't fully implemented WC yet). 
As you mentioned would be nice to have an example of WC with Algorand on the official WC repository to make this process easier for new projects that may appear in Algorand. That is why in [Gunny Games](https://www.gunnygames.com/) we want to share the code to make it publicly available.

It is necessary to mention that not only changes have to be made in "_Chain.cs_" but also in "_Dapp.cs_". 

`Chain.cs`
```

public static readonly string AvmNamespace = "algorand"; //AVM stands for Algorand Virtual Machine

public static readonly Chain AlgorandMainnet = new(AvmNamespace, "wGHE2Pwdvd7S12BL5FaOP20EGYesN73k", "Algorand MainNet", new Color(0.91f, 0f, 0f));

public static readonly Chain AlgorandTestnet = new(AvmNamespace, "SGO1GKSzyE7IEPItTxCByw9x8FmnrCDe", "Algorand TestNet", new Color(0.91f, 0f, 0f), true); //

public static readonly Chain[] All =
{
    AlgorandMainnet,
    AlgorandTestnet,
    Ethereum,
    EthereumGoerli,
    Optimism,
    OptimismGoerli,
    Polygon,
    Arbitrum,
    ArbitrumRinkeby,
    Celo,
    CeloAlfajores,
    xDai,
    Base
};
```

> **Note**: I did not make any changes to the logo or the color of the interface as I do not know what criteria you use. 

`Dapp.cs`

```
var requiredNamespaces = new RequiredNamespaces();

if (_selectedChains.Any(c => c.ChainNamespace == Chain.EvmNamespace))
{
    var eipChains = _selectedChains.Where(c => c.ChainNamespace == Chain.EvmNamespace);

    // TODO: make configurable
    var methods = new[]
    {
        "eth_sendTransaction",
        "personal_sign",
    };

    var events = new[]
    {
        "chainChanged", "accountsChanged"
    };

    var chainIds = eipChains.Select(c => c.FullChainId).ToArray();

    requiredNamespaces.Add(Chain.EvmNamespace, new ProposedNamespace()
    {
        Chains = chainIds,
        Events = events,
        Methods = methods
    });
}
else if (_selectedChains.Any(c => c.ChainNamespace == Chain.AvmNamespace))
{
    //ARC: Algorand Requests for Comments
    var arcChains = _selectedChains.Where(c => c.ChainNamespace == Chain.EvmNamespace);

    // TODO: make configurable
    var methods = new[]
    {
        "algo_signTxn"
    };

    var events = new[]
    {
        "accountChanged"
    };

    var chainIds = arcChains.Select(c => c.FullChainId).ToArray();

    requiredNamespaces.Add(Chain.EvmNamespace, new ProposedNamespace()
    {
        Chains = chainIds,
        Events = events,
        Methods = methods
    });
}
else
{
    throw new Exception("No EVM or AVM chains selected.");
}

return new ConnectOptions
{
    RequiredNamespaces = requiredNamespaces
};
```

_I hope this can help more people and soon the example can be published including Algorand in the list, if you need anything else from me I will be attentive to help._

**Reply by Dilosen** [reply]
Nice work on this.  I was able to implement it and get a wallet connection using a similar setup as yours. My issue is building out the wallet onto a mobile test, the QR codes did not generate on mobile but does generate in the editor, the API calls to the explorer pulled the corresponding networks and wallets.

Not sure if you're building for the mobile platform, but this is where I stopped last with it. 

Sidenote: did Pera add their wallet to the explorer? (Contacted them in the week to request WC wallet to be added to the explorer)

**Reply by sebas6612** [reply]
No, for the moment I'm just developing for Windows and Mac, it is in my plans to develop a mobile version but in the future.

**Reply by skibitsky** [reply]
Hey @sebas6612, this is awesome! Thank you for sharing the code. I will add it to the sample in the future updates.

**Comment by sebas6612** [comment]
> Hey @sebas6612, this is awesome! Thank you for sharing the code. I will add it to the sample in the future updates.

Hi @skibitsky, I see that you haven't added Algorand to the example yet, but I think it would be good to put it since the Algorand community is growing every day and it is a blockchain with ideal features for game development.

**Reply by skibitsky** [reply]
Hey, I didn't forget! However, it isn't a high priority at the moment. Also, I'm still now aware of any wallet in WalletConnect explorer that would support Algorand.

**Reply by Dilosen** [reply]
Hi @skibitsky - fair enough, we do have Pera Wallet support? Not sure, I haven't revisited this in a long while

**Reply by skibitsky** [reply]
Hey @moosesnWoop, AFAIK Pera Wallet supports connection via QR code, but for deep links they have to submit it to the WalletConnect Explorer.

**Reply by skibitsky** [reply]
Hey @sebas6612, the latest version of WalletConnect Modal sample now includes Algorand.

https://github.com/WalletConnect/WalletConnectUnity/blob/main/Packages/com.walletconnect.modal/Samples~/Modal%20Sample/Scripts/Dapp.cs#L78

**Reply by sebas6612** [reply]
Hi @skibitsky that's amazing!! thank you so much!!

**Comment by Dilosen** [comment]
@skibitsky lmfao they still haven't submitted to Wallet Connect? I've contacted them directly and they said they would! Goes to show.  Probably better off building for other chains right now.
